### PR TITLE
#163388140 Authenticated users should be able to comment on articles

### PR DESCRIPTION
**Pull request by PaulinhoNdegwa**
#### What does this PR do?
It enables an authenticated user to comment on articles and reply to comments

#### Description of Task to be completed?
- Created comment model and serializer
- Creates comment CRUD views
- Written comment tests and refactored

#### How should this be manually tested?
1. Clone the repo

```
$ git clone https://github.com/andela/ah-the-phoenix.git
```
1. Change it into the current directory

```
$ cd ah-the-phoenix
```
Checkout to the Feature branch

```
$ git checkout ft-articles-comments-163388140
```
Run the server
   `python manage.py runserver`
1. Comment on an article
      Send a `POST` request to  `http://127.0.0.1:8000/api/v1/articles/<slug>/comment/`
      Sample comment:
```
{
"body": "This is a comment"
}
```
image

2. GET all article comments
      Send a `GET` request to  `http://127.0.0.1:8000/api/v1/articles/<slug>/comment/`
image

3. GET one comment
    Send a `GET` request to  `http://127.0.0.1:8000/api/v1/articles/<slug>/comment/<comment_id>`
image

4. UPDATE a comment
    Send a `PUT` request to  `http://127.0.0.1:8000/api/v1/articles/<slug>/comment/<comment_id>`
image

5. DELETE a comment
    Send a `DELETE` request to  `http://127.0.0.1:8000/api/v1/articles/<slug>/comment/<comment_id>`
image

6. REPLY to a comment
    Send a `POST` request to  `http://127.0.0.1:8000/api/v1/articles/<slug>/comment/<comment_id>`
    _* You cannot reply to a reply_
image

#### Any background context you want to provide?
None

#### What are the relevant pivotal tracker stories?

 #163388140

### CODE REVIEW

File: authors/apps/articles/migrations/0001_initial.py [hunk 1]
```diff
@@ -1,14 +1,17 @@
-# Generated by Django 2.1.5 on 2019-02-14 18:54
+# Generated by Django 2.1.5 on 2019-02-18 16:07
```
Comment (Arusey) [inline authors/apps/articles/migrations/0001_initial.py]: Nice implementation. good job

File: authors/apps/articles/models.py [hunk 2]
```diff
@@ -41,3 +41,42 @@ def save(self, *args, **kwargs):
         if not self.slug:
             self.slug = self.generate_slug()
         super().save(*args, **kwargs)
+
+
+class Comments(models.Model):
```
Comment (sirghiny) [inline authors/apps/articles/models.py]: Model names should be singular.

File: authors/apps/articles/views.py [hunk 3]
```diff
@@ -85,3 +86,145 @@ def destroy(self, request, pk=None):
...
+    """
+    queryset = Comments.objects.all()
+    serializer_class = CommentSerializer
+    permission_classes = (IsAuthenticatedOrReadOnly,)
+    renderer_classes = (ArticleJsonRenderer,)
+
+    def get_specific_comment(self, article_id, comment_id, request):
+        """This methos a single comment related to a specific article"""
+        try:
+            Article.objects.get(pk=article_id)
```
Comment (sirghiny) [inline authors/apps/articles/views.py]: We'll need to abstract out the getting of common things e.g Articles in viewsets. You can leave it for now but change it in the code clean up.

Comment (PaulinhoNdegwa) [reply authors/apps/articles/views.py]: How do i abstract? Example in the articles scenario?

File: authors/apps/articles/views.py [hunk 4]
```diff
@@ -85,3 +86,145 @@ def destroy(self, request, pk=None):
...
+        except Exception:
+            return Response({"error": "Article does not exist"})
+        try:
+            comment = Comments.objects.filter(pk=comment_id,
+                                              article_id=article_id).first()
+        except Exception:
+            raise Http404("Data not found")
```
Comment (asirvex) [inline authors/apps/articles/views.py]: give a more specific error message

File: authors/apps/articles/views.py (hunk 3 above)
Comment (asirvex) [reply authors/apps/articles/views.py]: You can abstract the process of getting an article and returning an error if it is not found. In the latest changes there is a function called get_article that can do that

Comment (PaulinhoNdegwa) [reply authors/apps/articles/views.py]: Do you mean the `get_article` in the serializer?

File: authors/apps/articles/views.py (hunk 4 above)
Comment (PaulinhoNdegwa) [reply authors/apps/articles/views.py]: Done

File: authors/apps/articles/views.py (hunk 3 above)
Comment (PaulinhoNdegwa) [reply authors/apps/articles/views.py]: I will look into it

File: authors/apps/articles/models.py (hunk 2 above)
Comment (PaulinhoNdegwa) [reply authors/apps/articles/models.py]: Done

**Review by Arusey — COMMENTED** [review]
Looks good. Nice implementation

File: authors/apps/articles/migrations/0001_initial.py (hunk 1 above)
Comment (PaulinhoNdegwa) [reply authors/apps/articles/migrations/0001_initial.py]: Thanks.

File: authors/apps/articles/views.py (hunk 3 above)
Comment (PaulinhoNdegwa) [reply authors/apps/articles/views.py]: I have made the changes on the final rebase commit..

**Comment by dev-jey** [comment]
Great job on the threading aspect of the comments@PaulinhoNdegwa

**Comment by PaulinhoNdegwa** [comment]
> Great job on the threading aspect of the comments@PaulinhoNdegwa

Thanks @codact

**Comment by PaulinhoNdegwa** [comment]
> Looks good. Nice implementation

Hello @Arusey, Kindly be more specific on what on I accomplished well. Thanks!

**Pull request merged**